src/System.Text.JsonLab/System/Text/Json/JsonWriter.cs

```diff
@@ -145,8 +531,23 @@ public void WriteAttribute(string name, string value)
         /// <param name="value">The signed integer value to be written to JSON data.</param>
         public void WriteAttribute(string name, long value)
         {
-            WriteStartAttribute(name);
-            WriteNumber(value);
+            if (UseFastUtf8)
+            {
+                ReadOnlySpan<byte> nameSpan = MemoryMarshal.AsBytes(name.AsSpan());
+                int bytesNeeded = CalculateStartAttributeBytesNeeded(nameSpan, sizeof(byte));
```

Comment:
Is it faster to calculate and then write? Why not be optimistic and use TryWrite pattern?

Reply:
How would we figure out when to return false when using the TryWrite pattern? The only way I can think of is to call GetSpan() and compare it's length to how much we need (repeatedly). We have to compute bytes needed (or some upper bound) and make at least one call to GetSpan to do that anyway, and those calls are expensive (interface dispatch). Therefore, to make sure we make only one call to GetSpan and get at least what we need, we have to pre-compute the bytes needed (which is a relatively cheap operation). The alternative is several calls to GetSpan (which is what we originally had), or over-asking unnecessarily.

>  In most cases the buffer is large enough

That behavior is already built into the GetSpan implementation in ArrayFormatter, where we just return back the FreeSpan (without doing the expensive resize operation) when we have enough space.

The same issue applies with Advance, which is another benefit of pre-computing. We reduce the calls to Advance to just one, since we know exactly how much we need to move, ahead of time.

Reply:
It depends what you will be doing with this buffer. ILet's say you will be copying static data into it: use TryCopy. Or formar and int, use TryFormat, etc. i.e. turtles all the way down (ok, maybe all the way to the indexer setting a value in the span).

Reply:
Interesting, I tried the optimistic approach, and for methods where pre-computing is expensive, we do get a decent win! I will file an issue for this to apply "optimistic writing" to specific methods where pre-computing is non-trivial. We save ~15% for instance, for writing a string attribute (`WriteAttribute(string, string)`), where calculating the bytes needed requires a call to the encoder.

But, for methods like `WriteValue(long)`, where pre-computing the bytes needed is trivial (and we have to call DigitCount anyway), it is worse to use the TryWrite pattern (~20% slower).

I am not 100% certain that these savings from the TryWrite pattern can successfully offset the cost to re-do the work when we have to expand the buffer. They probably can, if we apply it selectively.